File: java/src/main/java/com/equities/bench/Workload.java
package com.equities.bench;

public final class Workload {
    public final byte[] op;
    public final int[] orderId;
    public final int[] symbol;
    public final byte[] side;
    public final int[] price;
    public final long[] qty;
    public final byte[] tif;
    public final int[] account;
    public final int nOps;
    public final int nSymbols;
    public final int nAccounts;

    private Workload(int nOps, int nSymbols, int nAccounts) {
        this.nOps = nOps;
        this.nSymbols = nSymbols;
        this.nAccounts = nAccounts;
        this.op = new byte[nOps];
        this.orderId = new int[nOps];
        this.symbol = new int[nOps];
        this.side = new byte[nOps];
        this.price = new int[nOps];
        this.qty = new long[nOps];
        this.tif = new byte[nOps];
        this.account = new int[nOps];
    }

    public static Workload generate(long seed, int nOps, int nSymbols, int nAccounts) {
        Workload w = new Workload(nOps, nSymbols, nAccounts);
        XorShift64 rng = new XorShift64(seed);
        int nextId = 1;
        for (int i = 0; i < nOps; i++) {
            int r = rng.nextBounded(100);
            if (r < 70 || nextId == 1) {
                w.op[i] = C.OP_ADD;
                w.orderId[i] = nextId++;
                w.symbol[i] = rng.nextBounded(nSymbols);
                w.side[i] = (byte) rng.nextBounded(2);
                w.price[i] = C.PRICE_MID + rng.nextBounded(C.PRICE_SPAN) - C.PRICE_SPAN / 2;
                w.qty[i] = 1 + rng.nextBounded(100);
                w.tif[i] = rng.nextBounded(10) == 0 ? C.TIF_IOC : C.TIF_GTC;
                w.account[i] = rng.nextBounded(nAccounts);
            } else if (r < 90) {
                w.op[i] = C.OP_CANCEL;
                w.orderId[i] = 1 + rng.nextBounded(nextId - 1);
                w.account[i] = rng.nextBounded(nAccounts);
            } else {
                w.op[i] = C.OP_MARKET;
                w.orderId[i] = nextId++;
                w.symbol[i] = rng.nextBounded(nSymbols);
                w.side[i] = (byte) rng.nextBounded(2);
                w.price[i] = 0;
                w.qty[i] = 1 + rng.nextBounded(100);
                w.tif[i] = C.TIF_IOC;
                w.account[i] = rng.nextBounded(nAccounts);
            }
        }
        return w;
    }
}

File: python/src/equities_bench/workload.py
from . import constants as c
from .rng import XorShift64


class Workload:
    def __init__(self, n_ops: int) -> None:
        self.op = [0] * n_ops
        self.order_id = [0] * n_ops
        self.symbol = [0] * n_ops
        self.side = [0] * n_ops
        self.price = [0] * n_ops
        self.qty = [0] * n_ops
        self.tif = [0] * n_ops
        self.account = [0] * n_ops
        self.n_ops = n_ops

    @staticmethod
    def generate(seed: int, n_ops: int, n_symbols: int, n_accounts: int) -> "Workload":
        w = Workload(n_ops)
        rng = XorShift64(seed)
        next_id = 1
        for i in range(n_ops):
            r = rng.next_bounded(100)
            if r < 70 or next_id == 1:
                w.op[i] = c.OP_ADD
                w.order_id[i] = next_id
                next_id += 1
                w.symbol[i] = rng.next_bounded(n_symbols)
                w.side[i] = rng.next_bounded(2)
                w.price[i] = c.PRICE_MID + rng.next_bounded(c.PRICE_SPAN) - c.PRICE_SPAN // 2
                w.qty[i] = 1 + rng.next_bounded(100)
                w.tif[i] = c.TIF_IOC if rng.next_bounded(10) == 0 else c.TIF_GTC
                w.account[i] = rng.next_bounded(n_accounts)
            elif r < 90:
                w.op[i] = c.OP_CANCEL
                w.order_id[i] = 1 + rng.next_bounded(next_id - 1)
                w.account[i] = rng.next_bounded(n_accounts)
            else:
                w.op[i] = c.OP_MARKET
                w.order_id[i] = next_id
                next_id += 1
                w.symbol[i] = rng.next_bounded(n_symbols)
                w.side[i] = rng.next_bounded(2)
                w.price[i] = 0
                w.qty[i] = 1 + rng.next_bounded(100)
                w.tif[i] = c.TIF_IOC
                w.account[i] = rng.next_bounded(n_accounts)
        return w
